FIX: fix cleanup warnings for errorbar timeseries

ivanovmg opened the pull request:
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Currently in ``pandas/tests/plotting/tests_frame.py`` there are several warnings emitted.

```
pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_mpl2_color_cycle_str
  /workspaces/pandas/pandas/plotting/_matplotlib/style.py:64: MatplotlibDeprecationWarning: Support for uppercase single-letter colors is deprecated since Matplotlib 3.1 and will be removed in 3.3; please use lowercase instead.
    [conv.to_rgba(c) for c in colors]

pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_errorbar_plot
pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_errorbar_plot
pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_errorbar_plot
pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_errorbar_timeseries
pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_errorbar_timeseries
pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_errorbar_timeseries
  /workspaces/pandas/pandas/plotting/_matplotlib/__init__.py:61: UserWarning: To output multiple subplots, the figure containing the passed axes is being cleared
    plot_obj.generate()

-- Docs: https://docs.pytest.org/en/stable/warnings.html
```

This PR handles warnings in ``test_errorbar_timeseries``.
If this approach is considered reasonable by the reviewers, then I will do the same for ``test_errorbar_plot`` as well as other similar warnings among the plotting-related tests.

ivanovmg committed:
FIX: fix cleanup warnings for errorbar timeseries

jreback commented:
so the test itself was causing the warnings?

jreback commented:
cc @charlesdong1991

ivanovmg commented:
> 
> 
> so the test itself was causing the warnings?

Yes, the test itself was causing the warning.

charlesdong1991 commented:
this reminds me of this https://github.com/pandas-dev/pandas/pull/13188 issue, the warning looks the same, should we use `with tm.assert_produces_warning(UserWarning):` to catch the part where `subplots=True`?

charlesdong1991 commented on pandas/tests/plotting/test_frame.py:
```diff
@@ -2765,35 +2765,62 @@ def test_errorbar_with_partial_columns(self):
+        # subplots inside, which led to uncaught warnings like this:
+        # UserWarning: To output multiple subplots,
+        # the figure containing the passed axes is being cleared
+        tdf.plot(kind=kind, yerr=tdf_err, subplots=True)
```
emm, does this actually do anything?

charlesdong1991 commented on pandas/tests/plotting/test_frame.py:
```diff
@@ -2765,35 +2765,62 @@ def test_errorbar_with_partial_columns(self):
+        # subplots inside, which led to uncaught warnings like this:
+        # UserWarning: To output multiple subplots,
+        # the figure containing the passed axes is being cleared
+        tdf.plot(kind=kind, yerr=tdf_err, subplots=True)
```
probably to add `plt.clf()` above to clear the figure?

ivanovmg replied:
I guess it is not required, because ``df.plot`` would create new figure anyway.

Here is an example.

```
>>> import pandas as pd
>>> df = pd.DataFrame({'a': [1, 2, 3, 4, 5]})
>>> import matplotlib.pyplot as plt
>>> fig = plt.figure()
>>> df.plot()
<matplotlib.axes._subplots.AxesSubplot object at 0x14512EB0>
>>> fig2 = plt.gcf()
>>> fig == fig2
False
>>> fig
<Figure size 640x480 with 0 Axes>
>>> fig2
<Figure size 640x480 with 1 Axes>
>>>
```

Do I understand it correctly?

charlesdong1991 replied:
yeah!

ivanovmg commented:
> 
> 
> this reminds me of this #13188 issue, the warning looks the same, should we use `with tm.assert_produces_warning(UserWarning):` to catch the part where `subplots=True`? or we should also use the same solution here to replace these warnings?

I did not use ``tm.assert_produces_warning`` because there is no way to match warning message this way.
I believe it is reasonable to match the message here.
What would others think?

charlesdong1991 commented:
true, I would suggest to use `tm.assert_produce_warning` here to align with other existing similar issue, and see if we could add an arg for `tm.assert_produce_warning` to also match the warning message (I haven't looked at how this function is structured, not sure if it's easy or not to add such match), and then do a small refactor to match the warning message in these places.

does it sound good to you?

ivanovmg commented:
> 
> 
> true, I would suggest to use `tm.assert_produce_warning` here to align with other existing similar issue, and see if we could add an arg for `tm.assert_produce_warning` to also match the warning message (I haven't looked at how this function is structured, not sure if it's easy or not to add such match), and then do a small refactor to match the warning message in these places.
> 
> does it sound good to you?

Absolutely, I will refactor to ``tm.assert_produces_warning`` and add a comment what kind of warning message is expected.
Then in a follow-on PR I can take a look at the extension of ``tm.assert_produces_warning`` to match messages.
I believe that extension of `tm.assert_produces_warning`` goes beyond the scope of the present PR.

charlesdong1991 commented:
yep! sounds good! 

you could use use `tm.assert_produce_warning` in this PR and add a comment to refer to the original 13188, so later could easily catch them! Please also do it for `test_errorbar_plot` test as well!

> I believe that extension of tm.assert_produces_warning goes beyond the scope of the present PR.

yeah, a follow-up would be great!

ivanovmg committed:
REF: refactor using assert_produces_warning

As agreed with charlesdong1991,
use assert_produces_warning
and leave a comment on which warning message is emitted

ivanovmg committed:
REF: catch warning in test_errorbar_plot

ivanovmg committed:
REF: extract test_errorbar_plot_different_kinds

Extract test_errorbar_plot_different_kinds
from test_errorbar_plot and parametrize it.

ivanovmg commented:
Did that. But also I decided to move from ``test_errorbar_plot`` part related to various kinds in a separate parametrize test. Can revert it if this is not suitable. But to me it seems that in general it would be required to split some test functions into smaller ones with more relevant names.

charlesdong1991 commented:
> I decided to move from test_errorbar_plot part related to various kinds in a separate parametrize test

sure!

charlesdong1991 reviewed (approved):
very nice!

could you double check that after this change, the warnings you post in the PR description are gone?

cc @jreback

charlesdong1991 commented on pandas/tests/plotting/test_frame.py:
```diff
@@ -2765,35 +2787,44 @@ def test_errorbar_with_partial_columns(self):
+            # UserWarning: To output multiple subplots,
+            # the figure containing the passed axes is being cleared
+            # Similar warnings were observed in GH #13188
+            _check_plot_works(tdf.plot, kind=kind, yerr=tdf_err, subplots=True)
+            fig = plt.gcf()
+            axes = fig.get_axes()
+            for ax in axes:
```
sorry, just noticed this, a minor question:

i think with `tm.assert_produce_warning`, we could still use the current way to check errorbars, right? `_check_has_errorbars` i recall can take in `axes/list-like` input.

```python
axes = _check_plot_works(	
                    tdf.plot, kind=kind, yerr=tdf_err, subplots=True	
                )
self._check_has_errorbars(axes, xerr=0, yerr=1)
```

ivanovmg replied:
Oh, yes, it does accept list-like. I will change to the original implementation.

ivanovmg commented:
I checked.
Here is the output.

```
(base) root@f31dd4516194:/workspaces/pandas# pytest pandas/tests/plotting/test_frame.py
======================================================================================================================== test session starts =========================================================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /workspaces/pandas, configfile: setup.cfg
plugins: cov-2.10.0, forked-1.2.0, xdist-1.34.0, hypothesis-5.23.11, asyncio-0.12.0
collected 191 items                                                                                                                                                                                                                                                  

pandas/tests/plotting/test_frame.py ...........................x....................................................................................................x..............................................................                            [100%]

========================================================================================================================== warnings summary ==========================================================================================================================
pandas/tests/plotting/test_frame.py::TestDataFramePlots::test_mpl2_color_cycle_str
  /workspaces/pandas/pandas/plotting/_matplotlib/style.py:64: MatplotlibDeprecationWarning: Support for uppercase single-letter colors is deprecated since Matplotlib 3.1 and will be removed in 3.3; please use lowercase instead.
    [conv.to_rgba(c) for c in colors]

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================================================================================================== 189 passed, 2 xfailed, 1 warning in 64.14s (0:01:04) ========================================================================================================
(base) root@f31dd4516194:/workspaces/pandas#
```

Another warning is handled in https://github.com/pandas-dev/pandas/pull/36981

ivanovmg committed:
REF: check errorbars for list-like axes

ivanovmg committed:
FIX: fix comments

jreback commented:
thanks @ivanovmg
